### PR TITLE
Feat: FlashmintNAV

### PR DESCRIPTION
**Pull request by edkim**

### CODE REVIEW

File: contracts/exchangeIssuance/FlashMintNAV.sol [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+    /**
+    * Issues a minimum amount of SetTokens for an exact amount of ETH.
+    *
+    * @param _setToken              Address of the SetToken to be issued
+    * @param _minSetTokenAmount     Minimum amount of SetTokens to be issued
+    * @param _reserveAssetSwapData  Swap data to trade WETH for reserve asset
+    */
+    function issueSetFromExactETH(
+        ISetToken _setToken,
+        uint256 _minSetTokenAmount,
```
Comment (edkim) [inline contracts/exchangeIssuance/FlashMintNAV.sol:216]: @janndriessen double-checking that you're able to integrate with this issuance function that doesn't return an exact amount of SetTokens?

Comment (janndriessen) [reply contracts/exchangeIssuance/FlashMintNAV.sol:216]: @edkim so we'd have to call `getIssueAmount` or `getRedeemAmountOut` to determine this `minSetTokenAmount`?

File: contracts/exchangeIssuance/FlashMintNAV.sol [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+    * @param _inputTokenAmount      Exact amount of input token to spend
+    * @param _reserveAssetSwapData  Swap data to trade input token for reserve asset. Use empty swap data if input token is the reserve asset.
+    * 
+    * @return                       Amount of SetTokens expected to be issued
+    */
+    function getIssueAmount(
+        ISetToken _setToken,
+        address _inputToken,
+        uint256 _inputTokenAmount,
+        DEXAdapterV2.SwapData memory _reserveAssetSwapData
```
Comment (janndriessen) [inline contracts/exchangeIssuance/FlashMintNAV.sol:160]: @edkim reserve asset would be e.g. USDC? and then the regular swap data format e.g. if inputToken DAI for DAI-USDC for the given `inputTokenAmount`?

File: test/integration/ethereum/flashMintNAV.spec.ts [hunk 3]
```diff
@@ -0,0 +1,548 @@
...
+        }
+
+        it("can estimate the amount of SetToken issued for a given amount of ETH", async () => {
+          const setTokenOutEstimate = await subjectQuote();
+          expect(setTokenOutEstimate).to.eq(BigNumber.from("25460056235206711599"));
+        });
+
+        it("should issue SetToken with ETH", async () => {
+          const setTokenOutEstimate = await subjectQuote();
+          subjectMinSetTokenOut = setTokenOutEstimate.mul(995).div(1000); // 0.5% slippage
```
Comment (edkim) [inline test/integration/ethereum/flashMintNAV.spec.ts:257]: @janndriessen here's an example of getting the minSetTokenAmount for issuance

File: test/integration/ethereum/flashMintNAV.spec.ts [hunk 4]
```diff
@@ -0,0 +1,548 @@
...
+
+        it("can estimate the amount of output token received for redeeming a given amount of Set Token", async () => {
+          const outputTokenAmount = await subjectQuote();
+          expect(outputTokenAmount).to.eq(BigNumber.from("881862431628006214"));
+        });
+
+        it("should redeem SetToken for ETH", async () => {
+          const outputAmountEstimate = await subjectQuote();
+          subjectMinOutputTokenAmount = outputAmountEstimate.mul(995).div(1000); // 0.5% slippage
+
```
Comment (edkim) [inline test/integration/ethereum/flashMintNAV.spec.ts:435]: Example of estimating `minOutputTokenAmount` when redeeming.

File: contracts/exchangeIssuance/FlashMintNAV.sol (hunk 1 above)
Comment (edkim) [reply contracts/exchangeIssuance/FlashMintNAV.sol:216]: @janndriessen exactly. See example for issuance here https://github.com/IndexCoop/index-coop-smart-contracts/pull/186/files#r1768661768.

And for redemption https://github.com/IndexCoop/index-coop-smart-contracts/pull/186/files#r1768665399

File: contracts/exchangeIssuance/FlashMintNAV.sol (hunk 2 above)
Comment (edkim) [reply contracts/exchangeIssuance/FlashMintNAV.sol:160]: Yes, reserve asset is USDC for USDCY. So for issuing with DAI the swapData would look something like
`{
            exchange: Exchange.UniV3,
            fees: [100],
            path: [addresses.tokens.dai, addresses.tokens.USDC],
            pool: ADDRESS_ZERO,
};`

See example in test https://github.com/IndexCoop/index-coop-smart-contracts/blob/e21b456c9a436e5d073ea4ea8ac54bc9101accb8/test/integration/ethereum/flashMintNAV.spec.ts#L336

File: contracts/interfaces/INAVIssuanceHook.sol [hunk 5]
```diff
@@ -0,0 +1,39 @@
+/*
+    Copyright 2020 Set Labs Inc.
```
[the remark this answers is not in the document]
Comment (edkim) [reply contracts/interfaces/INAVIssuanceHook.sol:2]: Copied over from https://github.com/IndexCoop/index-protocol/blob/master/contracts/interfaces/INAVIssuanceHook.sol

File: contracts/exchangeIssuance/FlashMintNAV.sol [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+
+    /* ============ External Functions ============ */
+
+    /**
+     * Withdraw tokens to selected address if they end up in the contract
+     *
+     * @param _tokens    Addresses of tokens to withdraw, specifiy ETH_ADDRESS to withdraw ETH
+     * @param _to        Address to send the tokens to
+     */
+    function withdrawTokens(IERC20[] calldata _tokens, address payable _to) external onlyOwner payable {
```
[the remark this answers is not in the document]
Comment (edkim) [reply contracts/exchangeIssuance/FlashMintNAV.sol]: Changed in [472a6dd](https://github.com/IndexCoop/index-coop-smart-contracts/pull/186/commits/472a6dddbb206f24e200eb067f38cafff7afa0e7)

File: contracts/exchangeIssuance/FlashMintNAV.sol [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+    using SafeERC20 for ISetToken;
+
+    /* ============ Constants ============== */
+
+    // Placeholder address to identify ETH where it is treated as if it was an ERC20 token
+    address constant public ETH_ADDRESS = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+
+    /* ============ State Variables ============ */
+
+    address public immutable WETH;
```
[the remark this answers is not in the document]
Comment (edkim) [reply contracts/exchangeIssuance/FlashMintNAV.sol:61]: Fixed in [6c8ab06](https://github.com/IndexCoop/index-coop-smart-contracts/pull/186/commits/6c8ab067ffeaf87bbd00b30ac200b70563fb0e4a)

File: contracts/exchangeIssuance/FlashMintNAV.sol [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+/**
+ * @title FlashMintNAV
+ * @author Index Cooperative
+ * @notice Part of a family of FlashMint contracts that allows users to issue and redeem SetTokens with a single input/output token (ETH/ERC20).
+ * Allows the caller to combine a DEX swap and a SetToken issuance or redemption in a single transaction.
+ * Supports SetTokens that use a NAV Issuance Module, and does not require use of off-chain APIs for swap quotes.
+ * The SetToken must be configured with a reserve asset that has liquidity on the exchanges supported by the DEXAdapterV2 library.
+ *
+ * See the FlashMint SDK for integrating any FlashMint contract (https://github.com/IndexCoop/flash-mint-sdk).
+ */
```
Comment (pblivin0x) [inline contracts/exchangeIssuance/FlashMintNAV.sol:45]: 🔥

File: test/integration/ethereum/flashMintNAV.spec.ts [hunk 9]
```diff
@@ -0,0 +1,548 @@
...
+      await setV2Setup.priceOracle.addAdapter(preciseUnitOracle.address);
+      await setV2Setup.priceOracle.addPair(tokenAddresses.usdc, tokenAddresses.usdc, preciseUnitOracle.address);
+      await setV2Setup.priceOracle.addPair(tokenAddresses.aEthUSDC, tokenAddresses.usdc, preciseUnitOracle.address);
+      await setV2Setup.priceOracle.addPair(tokenAddresses.cUSDCv3, tokenAddresses.usdc, preciseUnitOracle.address);
+      await setV2Setup.priceOracle.addPair(tokenAddresses.aUSDC, tokenAddresses.usdc, preciseUnitOracle.address);
+      erc4626Oracle = await deployer.setV2.deployERC4626Oracle(
+        tokenAddresses.gtUSDC,
+        usdc(1),
+        "gtUSDC - USDC Calculated Oracle",
+      );
```
Comment (pblivin0x) [inline test/integration/ethereum/flashMintNAV.spec.ts:124]: noting this constructor's arguments may change soon https://github.com/IndexCoop/index-protocol/pull/49

**Review by pblivin0x — APPROVED** [review]
LGTM 🚢

**Pull request merged**